---
title: "Riboviz: a new user's guide"
author: "Flic Anderson"
date: "16/12/2021"
output: 
  ioslides_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.2/jquery.min.js"></script>

<script>
    $(document).ready(function() {
      $('slide:not(.title-slide, .backdrop, .segue)').append('<footer label=\"bit.ly/riboviz-new-users"></footer>');    
    })
</script>

<style>
  footer:after {
    content: attr(label);
    font-size: 12pt;
    position: absolute;
    bottom: 20px;
    left: 700px;
    line-height: 1.9;
  }
</style>

<style type="text/css">
slides > slide:not(.nobackground):after {
  content: '';
}
</style>

<div class="notes">

Bio: 

Flic Anderson has been a Research Assistant in Bioinformatics in the Wallace Lab (within the Institute for Cell Biology) at the University of Edinburgh (July 2019 - December 2021). The Wallace Lab works to understand how translation works within cells (particularly of fungi) and what factors control and regulate protein synthesis. 

Flic is part of the development team for the 'riboviz' open source software package for processing and analysis of ribosome profiling data, working with colleagues from EPCC (University of Edinburgh) to make the codebase more robust & sustainable.   

With a background in Ecology and Botany (previously a Research Assistant at the Royal Botanic Garden Edinburgh), Flic has worked with a wide range of types of biological data using different software tools and programming languages (particularly R) for the last 8 years, and found herself drawn towards bioinformatics and research software engineering. 

Flic is a member of the Edinburgh Carpentries and a certified Carpentries instructor in foundational coding and data science skills for researchers. 


Title: Riboviz: a new user's guide


Abstract:  

Explaining how new users can get started with 'riboviz' open source software package for processing and analysing ribosome profiling (RNA sequence) data.

</div>



## HOW TO (RE-)USE THIS MATERIAL

This is a `.html` presentation created in `R Markdown` with `ioslides`. 

(It's been written in a [.Rmd](http://rmarkdown.rstudio.com) file, and I generated .html slides by 'knitting' it in Rstudio.)

You can check out the code used to make these slides at the [Talk repo](https://github.com/FlicAnderson/LabRetreat_20211108) on Github, and adapt it for your own presentations if you like - there's a MIT Licence on the repo, which means: 

*"Basically, you can do whatever you want as long as you include the original copyright and license notice in any copy of the software/source."*

Source: [tl;drLegal](https://tldrlegal.com/license/mit-license)


## Outline: 

* Riboviz 101
* Signs you might need to use riboviz
* Where to start
* How to install it 
* Assembling your input files
* How to configure riboviz
* How to run it
* The riboviz workflow
* How riboviz works
* How to check your results
* How to develop riboviz

These slides are available [on GitHub](https://github.com/FlicAnderson/20211216-LabTalk-riboviz)



## Riboviz 101

**Riboviz is open source software which processes & analyses ribosome profiling (sequencing) data**  

*Developed in collaborative project BBSRC-NSF funded project*

Ribosome profiling data helps unlock details of **translation**:   
_how the cell creates proteins from genetic code (DNA, RNA)._ 

riboviz helps researchers check **quality control** & understand the **characteristics** of the experimental data - ie. _do different experimental conditions change how the cell translates proteins from genes?_

<font size="2">riboviz at github: https://github.com/riboviz/riboviz</font>

<div class="notes">
...
</div>



## riboviz goals:

* end-to-end processing of Ribo-seq data
  - raw data through to useful figures
* flexible and robust
  - across library prep strategies (multiplexed, UMIs)
  - across organisms (bacteria, yeast, human, ...)
  - easy to extend
* transcript- or ORF-centric
  - splicing is a different problem
  - user chooses annotation relevant to their goals



## Signs you might need to use riboviz 

You want to:    

* **process and analyse ribo-seq data!**  
* install something that has been tested & works  
* avoid re-inventing the wheel   
* see the code (no black boxes)  
* make all your changes in 1 place  
* add to the software to answer specific research Qs   
* cite it*   

<font size="2">* Riboviz 2.0 pre-print: https://www.biorxiv.org/content/10.1101/2021.05.14.443910v1.full</font>



## Where to start

Skills you need:  

- **unix shell** / command line   
- **git** (& github) version control skills  
- basic *text editing* for the yaml configuration file  
- (basic *regular expressions* - optional, required if you have UMIs in your data!)  
- (*cluster computing* on Eddie - optional, but useful!)  



## How to install riboviz & dependencies

Easiest on Linux systems (possible to do from Windows/Mac, but developed/tested - in Ubuntu/CentOS linux).  

**["Install guide for riboviz & dependencies" in `docs/user/install.md` ](https://github.com/riboviz/riboviz/blob/main/docs/user/install.md)** 

**On Eddie compute cluster?**  Check and follow our documentation on setting up to run riboviz on Eddie at ["Running the riboviz workflow on Eddie" in `docs/user/run-on-eddie.md`](https://github.com/riboviz/riboviz/blob/develop/docs/user/run-on-eddie.md). Check with Edward that you've got access to `wallace_rna` lab group space for shared libraries.

**Containers:** install using docker/singularity - follow guides at the [riboviz containers repository](https://github.com/riboviz/containers)

**Developer dependencies:** for extra packages to get if you're developing or running tests look in [/docs/developer/install.md](https://github.com/riboviz/riboviz/blob/main/docs/developer/install.md)



## Assembling your input files

Check [example-datasets repository](https://github.com/riboviz/example-datasets) first, in case some of the files you need are already there...

Also check the [issue tickets](https://github.com/riboviz/example-datasets/issues) in case it's in progress already!

Require: 

* contaminants file `.fasta`
* annotation files `.fasta` & `.gff`
* sample reads file `.fastq`
* library prep / read structure information (e.g. adapters, UMIs, barcodes used etc.)  
* configuration file `.yaml`



## Assembling your input files

**Contaminants file:**  
- ribosomal rRNA  
- transfer RNA, snoRNA  
- other abundant non-coding RNAs  

Start with your favourite organism database or rRNA database. Ensembl, Silva rRNA database, FungiDB, VEUpathDB, Ensembl, BLAST, etc.   

Described and with examples available with `provenance.txt` files explaining how they were created for many taxa in [example-datasets](https://github.com/riboviz/example-datasets#contaminant-files). 



## Assembling your input files

**Annotation files:**  

* `.fasta` file of *transcript/extended-ORF sequences* (positive-strand only) AND `.gff` file giving *CDS/ORF/gene-feature position* within the matching `.fasta` file.  
* *Transcriptome-centric* annotation rather than genome annotation to reduce problems within analysis caused by splicing and antisense/opposite-strand reads. 
* Instructions on *generating* transcript-centric annotation files in [How to set up a new dataset](https://github.com/riboviz/example-datasets/blob/main/add-new-dataset.md#setting-up-a-new-species), including details on a script for this in `create_riboviz_style_cds_gff_acope3-278` branch of riboviz.  
* Check the discussions on riboviz issue ticket [#278 - Creating riboviz-friendly fasta and gff files](https://github.com/riboviz/riboviz/issues/278). 



## Assembling your input files

STOP: Are your annotation FASTA and GFF files riboviz-friendly? 

Follow ["Check FASTA and GFF files for coding sequence (CDS) features" at `docs/user/check-fasta-gff.md`](https://github.com/riboviz/riboviz/blob/main/docs/user/check-fasta-gff.md) and run the riboviz tool `check_fasta_gff` to check. 

* If it reports very few issues, ie. only tens, of which most can be traced back to biologically-interesting quirks with some googling and investigation of those genes, GOOD.  
* If the issue summary shows lots of issues, consider the quality of your transcriptome files or consider generating your own if you haven't already. 
* If ~half of the reads show issues relating to missing start/stop codons, this might indicate issues with 'strandedness'.  



## Assembling your input files

**Sample Reads files:**   

* `.fastq` or `.fastq.gz` format files.   
* If data is publicly available, download them using tools such as the `fasterq-dump` utility from SRA Toolkit. Details of [how to use this tool on Eddie](https://github.com/riboviz/riboviz/blob/main/docs/user/run-on-eddie.md#download-fastq-data-files-from-the-short-read-archive-sra-initial-setup), but if you're working locally, you may have to find system-specific details. 
* .fastq files are NOT found on example-datasets repository, as they are often several gigabytes large.



## Assembling your input files

**Configuration file:**  

- Check out [this repo](https://github.com/riboviz/generate-yaml) for a [shiny app](https://riboviz.shinyapps.io/generate-yaml/) that helps you generate a configuration file by Graphical User Interface.   
- the [example-datasets repository](https://github.com/riboviz/example-datasets) holds configuration files for public datasets published from a wide variety of species & annotation/contaminants files required for running them.   
- View the example files for the small test vignette datasets in [`vignette/`](https://github.com/riboviz/riboviz/tree/main/vignette) folder.  
- if dataset has *barcodes*, look up [parameters](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-config.md#configuration-parameters) `multiplex_fq_files` and `sample_sheet`. For *UMIs*, pay attention to UMI-specific config parameters `dedup_stats`, `dedup_umis`, `extract_umis`, `group_umis` and `umi_regexp`.



## How to configure riboviz: 

Go to **["Configuring the RiboViz workflow"](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-config.md)** for details on *riboviz input files* & how to *configure* riboviz by creating and editing your configuration file.

**Environment tokens:**   riboviz has a set of [configurable environment variable tokens](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-run-nextflow.md#defining-values-for-environment-variables) you can set to store path information about file locations.  There's an [example vignette](https://github.com/riboviz/riboviz/blob/develop/vignette/riboviz_env.yaml) using environmental tokens to demonstrate how to use them in a config file. This [environment variables example walk-through](https://github.com/riboviz/riboviz/blob/main/docs/user/run-env-vars-example.md) is also helpful in understanding how they work in practice.

**Handling UMIs & Barcodes:** There are some small simulated datasets demonstrating umi extraction, deduplication and demultiplexing features - find out how to find and run them [here](https://github.com/riboviz/riboviz/blob/main/docs/user/run-dedup-demultiplex-examples.md)



## How to configure riboviz further: 

**System resources:** a [config parameter](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-config.md#configuration-parameters) to be particularly aware of is `num_processes` - this adjusts how many processors will be used by parts of the workflow. When running on Eddie, ensure that `job_num_cpus` isn't greater than `num_processes`.

**Configuring on Eddie cluster**: for how to adjust Eddie job system resources, and how to generate a job script from a config file, check ["Job submission" at docs/user/run-on-Eddie.md](https://github.com/riboviz/riboviz/blob/main/docs/user/run-on-eddie.md#job-submission). All parameters starting `job_` are related to Eddie configuration.

**Upgrading old config files** [this guide](https://github.com/riboviz/riboviz/blob/main/docs/user/upgrade-config.md) details how to run the riboviz `upgrade_config_file` tool on your old config files to get new parameters from the latest version of riboviz. 




## How to run riboviz

**Run the vignette:** test your setup with short test 'vignette' - [`docs/user/run-vignette.md`](https://github.com/riboviz/riboviz/blob/main/docs/user/run-vignette.md) 

**Validate your run:** [`--validate-only`](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-run-nextflow.md#validate-configuration) flag lets you check paths, config file and key parameters are valid; highlight potential issues quickly.  

**Downsample your dataset:** use [this riboviz tool](https://github.com/riboviz/riboviz/blob/main/riboviz/tools/subsample_bioseqfile.py) to subsample fastq files for quicker testing/failing.

**Run a full dataset:** When your configuration file is set up & ready to go, [run the full dataset](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-run-nextflow.md#run-the-nextflow-workflow).

**Resume running after a failure:** if the run failed and you want to re-try, or have added a new sample, you can use the [`-resume`](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-run-nextflow.md#run-the-nextflow-workflow) parameter to start from the last successful point. 



## Riboviz Workflow: Inputs

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_inputs.png")
``` 
</div>

<div class="notes"> 

Additional organism-specific data files: ie the pipeline can run without them, but will not produce specific plots without them. 

</div> 


## Riboviz Workflow: Processing

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_processing-aligned.png")
``` 
</div>

<div class="notes"> 

Processing:  
* Removing adapter sequences  
* Remove UMIs (Unique Molecular Identifiers) & barcodes if present  
* Demultiplex / Deduplicate reads if required  
* Need to filter out contaminant reads  
* Align reads to transcriptome   

</div>



## Riboviz Workflow: Processing

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_processing-ribogrid.png")
``` 
</div>

<div class="notes">

Storing data in really efficient .h5 format (hierarchical, like a file system w/ folders type of organisation)

</div>


## Riboviz Workflow: Analysis

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_analysis.png")
``` 
</div>

<div class="notes">

Analyse & quantify data:   
Create outputs (including for quality-control, further analysis)

One main R script which generates these plots & various analysis data; this is run for each sample, outputting .tsv and .pdf files into a folder for each sample within the overall dataset output folder.

</div>



## Riboviz Workflow: + Visualization

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_visualization.png")
``` 
</div>

<div class="notes">

As part of making riboviz more user-friendly, wanted to generate a per-sample .html report based on the summary data and pdf outputs already generated by riboviz

</div>  



## Riboviz Workflow: ++ Visualization

<div class="centered">
```{r, out.width = "500"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/20211216-LabTalk-riboviz/main/imgs/riboviz_interactive_viz_vignette.png")
``` 
</div>
Vignette example, read lengths plot (both samples)

<div class="notes">

Can generate an R-shiny application based interactive space to explore your output data, allowing you to select specific samples, output plots, etc. New features can be added to this. 

</div>  



## How riboviz works 

**Different steps:** the different components and processes in the workflow are described in ["What the riboviz workflow does" at docs/user/prep-riboviz-operation.md](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-operation.md). The relevant [configuration parameters](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-operation.md#process-ribosome-profiling-sample-data) for each step are also listed. 

**Work directory:** Nextflow generates a specific directory within a `/work` folder for each iteration of each process. This is described in a section within [docs/user/prep-riboviz-operation.md](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-operation.md#nextflow-work-directory). 





## How to check your results (successes)

**Finding Dataset Outputs:** learn more about the output files at [docs/user/riboviz_outputs.md](https://github.com/riboviz/riboviz/blob/main/docs/user/riboviz_outputs.md)

**Visualizations:** riboviz generates separate per-sample .pdf plots by default, but each sample will also have a `.html` report generated which collates the plots for easier viewing & sharing.

**Interactive Visualizations:** to use riboviz' R-shiny interactive dataset explorer, follow this guide at [docs/user/run-shiny-server-operation.md](https://github.com/riboviz/riboviz/blob/main/docs/user/run-shiny-server-operation.md) after running riboviz successfully. If running on Eddie, check out [riboviz issue #431](https://github.com/riboviz/riboviz/issues/431).

**Nextflow reports:** [configuration parameters](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-config.md#configuration-parameters) starting `nextflow_` will let you generate different reports for a dataset run. Using the `-with-report` command line option when running riboviz generates a particularly useful html report. 



## How to check your results (errors)

**log files:** If a failure occurs, check the hidden `.command.*` files in the step-specific work folders (find how to locate them and which ones to look in [here](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-operation.md#nextflow-work-directory)). `.command.log` is one of my faves when troubleshooting. 

**standard output:** this is the text printed to the terminal when riboviz runs. This helps you see which processes have completed or failed. Standard output text might be captured into a `.o` file on Eddie (described within ["Submitting jobs" in 'docs/user/run-on-eddie.md](https://github.com/riboviz/riboviz/blob/main/docs/user/run-on-eddie.md#submitting-jobs)

The `-ansi-log false` nextflow option (described [here](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-run-nextflow.md#run-the-nextflow-workflow)) shows step-specific hashes in the riboviz output & helps you locate the right directory path when troubleshooting.



## How to develop riboviz 

**Run the tests:** read ["Developing and running tests" at `docs/developer/testing.md`](https://github.com/riboviz/riboviz/blob/main/docs/developer/testing.md) run tests to check the riboviz workflow produces expected results. Details on how to develop new tests or add to existing ones will be found in `docs/developer/testing.md`, with updates to address [riboviz issue #443](https://github.com/riboviz/riboviz/issues/443)

**Debugging issues:** useful guide on how to debug R scripts from commandline and nextflow tricks & tips in [/docs/developer/debugging.md](https://github.com/riboviz/riboviz/blob/main/docs/developer/debugging.md)

**Learn more about the Nextflow workflow management system**:  
- check out the 'alpha' Carpentries-style [Nextflow workshop materials](https://carpentries-incubator.github.io/workflows-nextflow/) developed by Ed-DaSH  
- go to the Nextflow docs at [nextflow.io](https://www.nextflow.io/)



## How to develop riboviz 

**Riboviz Style Guide:** for tools in python and R to help check code style & details on how to keep riboviz stylish, check [`docs/developer/coding-style`](https://github.com/riboviz/riboviz/blob/main/docs/developer/coding-style.md)

**Riboviz git etiquette:** if contributing to the riboviz codebase, see the [git-branching guide](https://github.com/riboviz/riboviz/blob/main/docs/developer/git-branching-model.md) on how to name branches, what order to merge things, when to run tests, and more. 

Further guides: [a range of developer guides to assist contributors](https://github.com/riboviz/riboviz/tree/main/docs/developer)




## riboviz uses best practices

* descriptive not prescriptive
* standard file formats wherever possible (.txt, .bam, .fasta, .gff)
* all parameters are described in one config file
* tests ensure that riboviz works correctly
* scripted with a workflow management system 

<font size="2">Best Practices for Scientific Computing - Wilson et al., PLoS Biology, 2014:
https://doi.org/10.1371/journal.pbio.1001745</font>



## What riboviz does well

* it works
  - on many datasets from many species
  - for many people
* count reads and assess read frame on pre-annotated ORFs
* quality control figures are informative
* easy to add new data for same transcriptome/organism
* multiplexed data & UMIs work smoothly
* can trim 5' untemplated mismatches added by Superscript reverse transcriptases
* thoroughly documented!



## What riboviz does not (yet) do (well)

* overlapping ORFs
  - not tested on RNA viruses
* discover new ORFs
* rigorously deal with multimapping reads
* poor support for matched RNA-seq data
  - so export counts from riboviz for differential translation
* new organisms/annotations are getting better, but still need user-effort.

There are other tools.

We are also interested in extending riboviz for **your** needs:   
Put in a feature request!



## In development... 

"Riboviz Choose Your Own Adventure..."

*Q5: Are you familiar with the Eddie compute cluster?*  

- Yes: You're ready to proceed (Go to Q6)  
- No: The Eddie team have put together some documentation and an ['Introduction to Eddie' course](https://www.wiki.ed.ac.uk/display/ResearchServices/Introduction+to+Eddie) you can work through to familiarise yourself with the Eddie system. It will requite your unix shell skills, and a university login to access. 

**Coming to a riboviz repo near you soon!**


## Thanks/Acknowledgements

[riboviz](https://github.com/riboviz/riboviz) is a BBSRC-NSF funded collaborative project: 

* **The Wallace Lab** - *University of Edinburgh*:  
Edward Wallace, + past students (Ania Kurowska, Siyin Xue, Emma MacKenzie, Sophie Winterbourne)  
* **The Shah Lab** -  *Rutgers University*:  
Premal Shah, John Favate, Alex Cope, Peter Tilton  
* **The Lareau Lab** - *University of California, Berkeley*:  
Liana Lareau, Amanda Mok, Vikram Shivakumar  
* **EPCC** - *University of Edinburgh*:  
Kostas Kavousannakis, Mike Jackson, Junchen Liu  

<div class="notes">
... slide notes ...
</div>



# END.


# Extra Technical Info / Explanations

# How to adapt to your own dataset

<div class="notes">
focus on key considerations: 
 - annotation files (inc. contaminants) - checkfastagff tool
 - the annotation file follows the biological question and must be chosen by the user
 - read structure (adapters, UMI/regex)
 - everything goes in config.yaml
refer to documentation for config.yaml
explain example-datasets repository
</div>



## Read structure

- Read the methods section for adapters, barcodes, UMIs
- If you did the library prep, you know this
- [UMI-tools](https://github.com/CGATOxford/UMI-tools/) allows flexible descriptions of barcodes and UMIs
  - `(?P<umi_1>.{8})` extracts an 8nt UMI from the 5' end of the read. 
  - `^(?P<umi_1>.{4}).+(?P<umi_2>.{4})(?P<cell_1>.{3})$` extracts a 3nt barcode from the 3' end of a read then extracts a 4nt UMI from the 5' end and a 4nt UMI from the 3' end.
  
<div class="notes"> 
riboviz includes mini test datasets including UMIs and barcodes: https://github.com/riboviz/riboviz/blob/main/docs/user/run-dedup-demultiplex-examples.md
</div>



## Transcript data needed for new organisms

This is the most unpredictable part, sadly. Budget 1-2 weeks.

**EITHER transcriptome:** use a published genome annotation with UTRs and splicing to create stranded transcript sequences in fasta format.

*But*, these are available for few organisms, and usually have weird formatting problems.

**OR padded ORFeome:** For each open reading frame of interest, the (spliced) ORF with fixed-with genomic sequence either side...



## Pick the annotation relevant to your biological question

| transcriptome  | padded ORFeome  |
|---|---|
| more realistic  | easier to generate  |
| all mRNAs  | ORF translation only |
| UTRs, uORFs  | -  |



## How to check the annotation

Reality-check your fasta and gff match with `check_fasta_gff.py` included with riboviz:

* start codons
* stop codons
* sense codons (no internal stops)

This flags big problems, like reverse-strand reads or intron inclusion.

You can still run riboviz on an annotation with errors: it is descriptive not prescriptive.

<div class="notes"> 
For user documentation on how to do this, check here: https://github.com/riboviz/riboviz/blob/main/docs/user/check-fasta-gff.md
</div>

