---
title: "Riboviz: a new user's guide"
author: "Flic Anderson"
date: "16/12/2021"
output: 
  ioslides_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.2/jquery.min.js"></script>

<script>
    $(document).ready(function() {
      $('slide:not(.title-slide, .backdrop, .segue)').append('<footer label=\"bit.ly/riboviz-new-users"></footer>');    
    })
</script>

<style>
  footer:after {
    content: attr(label);
    font-size: 12pt;
    position: absolute;
    bottom: 20px;
    left: 700px;
    line-height: 1.9;
  }
</style>

<style type="text/css">
slides > slide:not(.nobackground):after {
  content: '';
}
</style>

<div class="notes">

Bio: 

Flic Anderson has been a Research Assistant in Bioinformatics in the Wallace Lab (within the Institute for Cell Biology) at the University of Edinburgh (July 2019 - December 2021). The Wallace Lab works to understand how translation works within cells (particularly of fungi) and what factors control and regulate protein synthesis. 

Flic is part of the development team for the 'riboviz' open source software package for processing and analysis of ribosome profiling data, working with colleagues from EPCC (University of Edinburgh) to make the codebase more robust & sustainable.   

With a background in Ecology and Botany (previously a Research Assistant at the Royal Botanic Garden Edinburgh), Flic has worked with a wide range of types of biological data using different software tools and programming languages (particularly R) for the last 8 years, and found herself drawn towards bioinformatics and research software engineering. 

Flic is a member of the Edinburgh Carpentries and a certified Carpentries instructor in foundational coding and data science skills for researchers. 


Title: Riboviz: a new user's guide


Abstract:  

Explaining how new users can get started with 'riboviz' open source software package for processing and analysing ribosome profiling (RNA sequence) data.

</div>



## HOW TO (RE-)USE THIS MATERIAL

This is a `.html` presentation created in `R Markdown` with `ioslides`. 

(It's been written in a [.Rmd](http://rmarkdown.rstudio.com) file, and I generated .html slides by 'knitting' it in Rstudio.)

You can check out the code used to make these slides at the [Talk repo](https://github.com/FlicAnderson/LabRetreat_20211108) on Github, and adapt it for your own presentations if you like - there's a MIT Licence on the repo, which means: 

*"Basically, you can do whatever you want as long as you include the original copyright and license notice in any copy of the software/source."*

Source: [tl;drLegal](https://tldrlegal.com/license/mit-license)


## Outline: 

* Signs you might need to use riboviz
* Where to start
* How to install it 
* How to run it
* How riboviz works
* How to know it's working 
* How to share your results

These slides are available [on GitHub](https://github.com/FlicAnderson/20211216-LabTalk-riboviz)



## Thanks/Acknowledgements

riboviz is a BBSRC-NSF funded collaborative project: 

* **The Wallace Lab** - *University of Edinburgh*:  
Edward Wallace, + past students (Ania Kurowska, Siyin Xue, Emma MacKenzie, Sophie Winterbourne)  
* **The Shah Lab** -  *Rutgers University*:  
Premal Shah, John Favate, Alex Cope, Peter Tilton  
* **The Lareau Lab** - *University of California, Berkeley*:  
Liana Lareau, Amanda Mok, Vikram Shivakumar  
* **EPCC** - *University of Edinburgh*:  
Kostas Kavousannakis, Mike Jackson, Junchen Liu  

<div class="notes">
... slide notes ...
</div>



## Signs you might need to use riboviz 

You want to:    

* **process and analyse ribo-seq data!**  
* install something that has been tested & works  
* avoid re-inventing the wheel   
* see the code (no black boxes)  
* make all your changes in 1 place  
* add to the software to answer specific research Qs   
* cite it*   

<font size="2">* Riboviz 2.0 pre-print: https://www.biorxiv.org/content/10.1101/2021.05.14.443910v1.full</font>



## Goals of riboviz

* end-to-end processing of Ribo-seq data
  - raw data through to useful figures
* flexible and robust
  - across library prep strategies (multiplexed, UMIs)
  - across organisms (bacteria, yeast, human, ...)
  - easy to extend
* transcript- or ORF-centric
  - splicing is a different problem
  - user chooses annotation relevant to their goals



## riboviz uses best practices

* descriptive not prescriptive
* standard file formats wherever possible (.txt, .bam, .fasta, .gff)
* all parameters are described in one config file
* tests ensure that riboviz works correctly
* scripted with a workflow management system (Nextflow - check out the 'alpha' Carpentries-style [Nextflow workshop](https://carpentries-incubator.github.io/workflows-nextflow/) material) 

<font size="2">Best Practices for Scientific Computing - Wilson et al., PLoS Biology, 2014:
https://doi.org/10.1371/journal.pbio.1001745</font>



## What riboviz does well

* it works
  - on many datasets from many species
  - for many people
* count reads and assess read frame on pre-annotated ORFs
* quality control figures are informative
* easy to add new data for same transcriptome/organism
* multiplexed data & UMIs work smoothly
* can trim 5' untemplated mismatches added by Superscript reverse transcriptases
* thoroughly documented



## Documentation  

Extensive documentation helps to:   
* get running with riboviz  
* configure and adapt riboviz for your dataset   
* assist with your specific research questions  

Key guides include:

* [configuring riboviz](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-config.md) 
* [detailed guide to how the riboviz workflow operates](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-operation.md)
* [how to run the vignette and check results](https://github.com/riboviz/riboviz/blob/main/docs/user/run-vignette.md)
* [step-by-step how to run the main workflow](https://github.com/riboviz/riboviz/blob/main/docs/user/prep-riboviz-run-nextflow.md) 
* [how to get started with a new dataset](https://github.com/riboviz/example-datasets/blob/main/add-new-dataset.md)
* [visualize your dataset interactively](https://github.com/riboviz/riboviz/blob/develop/docs/user/run-shiny-server-operation.md)
* [a range of developer guides to assist contributors](https://github.com/riboviz/riboviz/tree/main/docs/developer)



## What riboviz does not (yet) do (well)

* overlapping ORFs
  - not tested on RNA viruses
* discover new ORFs
* rigorously deal with multimapping reads
* poor support for matched RNA-seq data
  - so export counts from riboviz for differential translation
* new organisms/annotations are getting better, but still need user-effort.

There are other tools.

We are also interested in extending riboviz for **your** needs:   
Put in a feature request!



## Riboviz Workflow: Inputs

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_inputs.png")
``` 
</div>

<div class="notes"> 

Additional organism-specific data files: ie the pipeline can run without them, but will not produce specific plots without them. 

</div> 



## Riboviz Workflow: Processing

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_processing-aligned.png")
``` 
</div>

<div class="notes"> 

Processing:  
* Removing adapter sequences  
* Remove UMIs (Unique Molecular Identifiers) & barcodes if present  
* Demultiplex / Deduplicate reads if required  
* Need to filter out contaminant reads  
* Align reads to transcriptome   

</div>



## Riboviz Workflow: Processing

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_processing-ribogrid.png")
``` 
</div>

<div class="notes">

Storing data in really efficient .h5 format (hierarchical, like a file system w/ folders type of organisation)

</div>


## Riboviz Workflow: Analysis

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_analysis.png")
``` 
</div>

<div class="notes">

Analyse & quantify data:   
Create outputs (including for quality-control, further analysis)

One main R script which generates these plots & various analysis data; this is run for each sample, outputting .tsv and .pdf files into a folder for each sample within the overall dataset output folder.

</div>



## Riboviz Workflow: + Visualization

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_visualization.png")
``` 
</div>

<div class="notes">

As part of making riboviz more user-friendly, wanted to generate a per-sample .html report based on the summary data and pdf outputs already generated by riboviz

</div>  



## Riboviz Workflow: ++ Visualization

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_visualization.png")
``` 
</div>

https://github.com/riboviz/riboviz/blob/develop/docs/user/run-shiny-server-operation.md

<div class="notes">

Can generate an R-shiny application based interactive space to explore your output data, allowing you to select specific samples, output plots, etc. New features can be added to this. 

</div>  



## riboviz workflow by sample

* **cutadapt** trim adapters
* **hisat2** to contaminants, align & discard
* **hisat2** to transcriptome, align & keep
* **trim_5p_mismatch** remove untemplated 5' 
* **samtools** sort and index bam file
* **bedtools genomecov** output bedgraphs
* **bam_to_h5.R** count reads by 5' position and read length
* **generate_stats_figs.R** read frame and read counts by ORF

There are options, e.g. multiplexing, UMIs.



## Finding and interpreting riboviz output  

Where does the output go? 



<div class="notes">
 - talk through where the files are
 - show and talk through .html output from the new downsampled dataset.
 - have one example from another full-size dataset to compare?
</div>


# How to adapt to your own dataset

<div class="notes">
focus on key considerations: 
 - annotation files (inc. contaminants) - checkfastagff tool
 - the annotation file follows the biological question and must be chosen by the user
 - read structure (adapters, UMI/regex)
 - everything goes in config.yaml
refer to documentation for config.yaml
explain example-datasets repository
</div>



## Key ingredients for a new dataset

For each dataset:

- read structure / library prep
- the raw data as fastq files

For new organisms or strains:

- transcript sequences and CDS maps
- contaminant sequences

Build from the most relevant example dataset: 

https://github.com/riboviz/example-datasets



## example-datasets repository is where we add new datasets

This example-datasets repository is for the configuration files and genome/annotation files needed to run the riboviz ribosome profiling pipeline on specific datasets. 

It aims to:

- provide specific example datasets for new users to try or to adapt
- share up-to-date tested example datasets between the riboviz development team

https://github.com/riboviz/example-datasets



## Example datasets are collected by organism

https://github.com/riboviz/example-datasets




## Read structure

- Read the methods section for adapters, barcodes, UMIs
- If you did the library prep, you know this
- [UMI-tools](https://github.com/CGATOxford/UMI-tools/) allows flexible descriptions of barcodes and UMIs
  - `(?P<umi_1>.{8})` extracts an 8nt UMI from the 5' end of the read. 
  - `^(?P<umi_1>.{4}).+(?P<umi_2>.{4})(?P<cell_1>.{3})$` extracts a 3nt barcode from the 3' end of a read then extracts a 4nt UMI from the 5' end and a 4nt UMI from the 3' end.
  
<div class="notes"> 
riboviz includes mini test datasets including UMIs and barcodes: https://github.com/riboviz/riboviz/blob/main/docs/user/run-dedup-demultiplex-examples.md
</div>



## Transcript data needed for new organisms

This is the most unpredictable part, sadly. Budget 1-2 weeks.

**EITHER transcriptome:** use a published genome annotation with UTRs and splicing to create stranded transcript sequences in fasta format.

*But*, these are available for few organisms, and usually have weird formatting problems.

**OR padded ORFeome:** For each open reading frame of interest, the (spliced) ORF with fixed-with genomic sequence either side...



## Pick the annotation relevant to your biological question

| transcriptome  | padded ORFeome  |
|---|---|
| more realistic  | easier to generate  |
| all mRNAs  | ORF translation only |
| UTRs, uORFs  | -  |



## How to check the annotation

Reality-check your fasta and gff match with `check_fasta_gff.py` included with riboviz:

* start codons
* stop codons
* sense codons (no internal stops)

This flags big problems, like reverse-strand reads or intron inclusion.

You can still run riboviz on an annotation with errors: it is descriptive not prescriptive.

<div class="notes"> 
For user documentation on how to do this, check here: https://github.com/riboviz/riboviz/blob/main/docs/user/check-fasta-gff.md
</div>



## Contaminants

- ribosomal rRNA
- transfer RNA, snoRNA
- other abundant non-coding RNAs

Start with your favourite organism database or rRNA database.

Ensembl, Silva rRNA database, FungiDB, VEUpathDB, Ensembl, BLAST, ...









# Extra Technical Info / Explanations
